src/osp/id_registry.h

```diff
@@ -0,0 +1,349 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "hierarchical_bitset.h"
+
+#include <cassert>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace osp
+{
+
+template<typename TYPE_T>
+struct type_identity
+{
+    using type = TYPE_T;
+};
+
+template<typename TYPE_T>
+using type_identity_t = typename type_identity<TYPE_T>::type;
+
+template<typename TYPE_T, typename = void>
+struct underlying_int_type;
+
+template<typename TYPE_T>
+struct underlying_int_type< TYPE_T, std::enable_if_t< std::is_enum_v<TYPE_T> > >
+ : std::underlying_type<TYPE_T>
+{ };
+
+template<typename TYPE_T>
+struct underlying_int_type< TYPE_T, std::enable_if_t< ! std::is_enum_v<TYPE_T> > >
+ : type_identity<TYPE_T>
+{ };
+
+template<typename TYPE_T>
+using underlying_int_type_t = typename underlying_int_type<TYPE_T>::type;
+
+enum class Test : uint32_t {};
+static_assert(std::is_same_v<underlying_int_type_t<int>, int>);
+static_assert(std::is_same_v<underlying_int_type_t<Test>, uint32_t>);
+
+//-----------------------------------------------------------------------------
+
+
+template<class TYPE_T>
+constexpr TYPE_T id_null() noexcept
+{
+    using id_int_t = underlying_int_type_t<TYPE_T>;
+    return TYPE_T(std::numeric_limits<id_int_t>::max());
+}
```

Comment:
Not sure I understand the purpose of this song and dance, what are the static asserts on 63/64 for, and are they the primary purpose of all these nested `using`s?

Reply:
Huh, lines 63 and 64 is leftover test code I forgot to delete, or possibly move to a unit test.

So there's `std::underlying_type`, which can be used to get the underlying integer type from an enum type. However, passing an integer type into it will cause a compile error.

`underlying_int_type` fixes this by instead returning the same type, if given an integer.

I should probably put a comment explaining this.

Reply:
Ok makes sense, yeah I didn't grasp the enum aspect until now. A comment would help.